crates/bevy_ecs/src/world/mod.rs

```diff
@@ -1836,6 +1855,15 @@ impl World {
         }
     }
 
+    /// Applies the internal [`CommandQueue`] to self
+    #[inline]
+    pub fn flush_commands(&mut self) {
+        if !self.flushing_commands && !self.command_queue.is_empty() {
+            let mut commands = std::mem::take(&mut self.command_queue);
+            commands.apply(self);
+        }
```

Comment:
```suggestion
        if self.flushing_commands || self.command_queue.is_empty() {
            return;
        }
        let mut commands = std::mem::take(&mut self.command_queue);
        commands.apply(self);
        
        // Replace the command queue to reuse the allocation.
        if self.command_queue.buffer_len() < commands.buffer_len() {
            commands.append(&mut self.command_queue);
            std::mem::replace(commands, &mut self.command_queue);
        }
```

`CommandQueue::buffer_len` would have to be added.

Reply:
On second thought, the `append` is unnecessary since afaict the world's command queue will always be empty after `CommandQueue::apply` returns. Is that accurate?

Reply:
Yes `CommandQueue::apply` will continuously drain the world's command queue.

Reply:
Bikeshedding:

Tbh this behavior seems counterintuitive -- I wouldn't expect that calling `apply` on *any* `CommandQueue` would consume the world's command queue. I think of `CommandQueue` as being a lower-level building-block, but now it feels more entangled with the API of `World`. It might be more intuitive if we make `CommandQueue::apply` private, and expose a method on `World` for applying commands from a command queue.

Reply:
I agree, that API would be more coherent, I didn't see it as worth adding to the PR as it felt like unnecessary noise but I do think it's cleaner, happy to do it now or we can leave it to an easy follow up.

Reply:
What if instead of applying the world's command queue, we temporarily swap it out, apply other commands, then restore it. This would mirror how `World::last_change_tick` works, where each exclusive system is a scope with its own `last_change_tick` that swaps it out and replaces it. Here, each invocation of `CommandQueue::apply` would create a scope with a new world command queue, then restore the old one at the end.
Thoughts?

Reply:
I wasn't able to enforce that all commands are always flushed from, for example, `.insert` calls on `EntityWorldMut` as flushing them on drop breaks `into_unsafe_entity_cell` (I think that's what happened, it was a while back, it might be possible with more refactoring).

This means that a `.insert` call could trigger hooks and any resulting commands would be stuck in the world's command queue while other command buffers are applied. Since we don't make any guarantees about command ordering I don't think this is technically wrong but I would consider it a little undesirable. Always applying all commands would make the ordering more consistent.

Reply:
Hmm. I suppose it's better for commands to get applied in a surprising order than an incorrect order.

Reply:
We should call `World::flush_commands` at the end of `ExcluiveFunctionSystem::run` (probably also at the end of the default implementation for `System::run`).